lib/pbench/server/api/resources/query_apis/metadata_index/table_of_content.py

```diff
@@ -0,0 +1,95 @@
+from logging import Logger
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    JSON,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.api.resources.query_apis import CONTEXT
+from pbench.server.api.resources.query_apis.metadata_index import RunIdBase
+
+
+class TableOfContent(RunIdBase):
+    """
+    Get the names of controllers within a date range.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("run_id", ParamType.STRING, required=True),
+                Parameter("parent", ParamType.STRING, required=True),
+            ),
+        )
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        parent = json_data.get("parent")
+        run_id = context["run_id"]
+        dataset = context["dataset"]
+
+        self.logger.info(
+            "Discover Table of Content for run_id {} and directory {}", run_id, parent,
+        )
+
+        # Retrieve the ES indices that belong to this run_id from the metadata
+        # table
+
+        indices = self.get_index(dataset, "run-toc")
+        if not indices:
+            self.logger.debug(
+                f"Found no indices matching the prefix run-toc"
+                f"for a dataset {dataset!r}"
+            )
+            abort(HTTPStatus.NOT_FOUND, message="Found no matching indices")
+
+        return {
+            "path": f"/{indices}/_search",
+            "kwargs": {
+                "json": {
+                    "size": 100,
+                    "query": {
+                        "bool": {
+                            "filter": [
+                                {"match": {"run_data_parent": run_id}},
+                                {
+                                    "dis_max": {
+                                        "queries": [
+                                            {"term": {"directory": parent}},
+                                            {"regexp": {"directory": f"{parent}[^/]+"}},
+                                        ]
```

Comment:
Interesting point.

Neither the `parent` nor `directory` fields in the index end with a trailing slash, and we'd generally expect them to start with `"/"` to traverse the tree from the top so they'll get back a set of nested directory paths like `"/1-read-4KiB/sample1"` ... which when returned won't work in this expression.

On the other hand, I don't see why we shouldn't allow someone to just ask for `"/1-read-4KiB/sample1/"` ... so I think we probably ought to *remove* a trailing slash, if present, from the `json_data["parent"]`, and then add it in the `"regex"` query term (because we don't want it for the exact match term, either; though, note that, if we try to combine the two terms, we'll need to deal with the fact that the slash will be in the subdirectories but not in the "root" we're looking for, so that would need to be a conditional subexpression like `(/[^/]*)?`)

Reply:
I will experiment with this more.

Reply:
Having reexamined this, I'm now convinced that my previous construct was correct and that these two clauses _cannot_ be combined...so I'm resolving this conversation....